### PR TITLE
feat: Add Drop method to DefaultCtx for silent connection termination

### LINKED ISSUE #3252: 📝 [Proposal]: Add silent request termination
### Feature Proposal Description

This issue proposes to add a method to `fiber.Ctx` that allows users to silently terminate HTTP requests in Fiber without sending any response headers nor response body. It should be a new method for the [`Ctx` interface](https://github.com/gofiber/fiber/blob/main/ctx.go#L51).

This feature will be helpful for:
- DDOS protection (avoid responding to malicious traffic)
- Hiding endpoints (stop bots from scraping sensitive endpoints by ignoring unauthenticated requests)

> **Note:**  While these methods are achievable through using a reverse proxy (e.g. NGINX), providing a native way
> of doing this in Fiber will give more control to users of how to implement the above for their use case (e.g. a user's app is running on a low-resource machine).

This feature will work by closing the underlying context's `net.Conn` using the  `fasthttp.(*RequestCtx).Conn().Close()` method. Fiber provides access to the underlying `fasthttp.(*RequestCtx)` through `c.Context()`. Thank you, @guno1928, for helping me find the simplest way to do this on Fiber's Discord Server! Your help is greatly appreciated.

The method's name is currently undecided, but here are a few ideas:

- [ ] `c.End()` ~(look at **Alignment with Express API**)~ (`Drop()` is different from the Express equivalent `res.end()`)
- [ ] `c.Close()`
- [ ] `c.Terminate()`
- [ ] `c.Stop()`
- [x] `c.Drop()`

Please feel free to give to an opinion of which of these (or any other recommendations) would be a clear name for the method.

### Alignment with Express API

The Express API uses the function ~`response.connection.end()`~ `response.connection.destroy()` to silently terminate HTTP requests:

```js
app.get("/", (request, response) => {
  response.connection.destroy();
  return;
});
```

To be similar to Express.js, this feature proposes to use a single method for the same functionality, called `Drop()`.
```go
app.Get("/", func (c fiber.Ctx) error {
 return c.Drop()
})
```

### HTTP RFC Standards Compliance

N/A, this feature will not affect how Fiber complies with the HTTP RFC standards.

### API Stability

This feature will not require changing any existing methods in `fiber.Ctx`, so it should be safe to add. As fasthttp's interface is stable, there should be minimal changes or deprecations in the future.

### Feature Examples

```go
package main

import (
	"log"
	"net"

	"github.com/gofiber/fiber/v3"
)

func main() {
	app := fiber.New()
	app.Get("/", func (c fiber.Ctx) error {
		name := c.Query("name")
		// Silently terminate requests without name
		if name == "" {
			return c.Drop()
		}
		return c.SendString("Hello, " + name + "!")
	})
	log.Fatal(app.Listen(":3000"))
}
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have searched for existing issues that describe my proposal before opening this one.
- [X] I understand that a proposal that does not meet these guidelines may be closed without explanation.

### PR DESCRIPTION
**Pull request by ryanbekhen**
# Description

This pull request introduces the Drop method to the DefaultCtx structure. The Drop method allows developers to silently terminate client connections without sending any HTTP headers or response body. This functionality is particularly useful for scenarios such as:

- Mitigating DDoS attacks.
- Blocking access to sensitive endpoints.
- Enhancing resource management by closing unwanted connections.

This addition aligns with Fiber’s goal of providing high-performance web server functionalities with fine-grained control.

Fixes #3252 

## Changes introduced

- [ ] Benchmarks: Describe any performance benchmarks and improvements related to the changes.
- [ ] Documentation Update: Detail the updates made to the documentation and links to the changed files.
- [ ] Changelog/What's New: Include a summary of the additions for the upcoming release notes.
- [ ] Migration Guide: If necessary, provide a guide or steps for users to migrate their existing code to accommodate these changes.
- [ ] API Alignment with Express: Explain how the changes align with the Express API.
- [ ] API Longevity: Discuss the steps taken to ensure that the new or updated APIs are consistent and not prone to breaking changes.
- [ ] Examples: Provide examples demonstrating the new features or changes in action.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (improvement to existing features and functionality)
- [ ] Documentation update (changes to documentation)
- [ ] Performance improvement (non-breaking change which improves efficiency)
- [ ] Code consistency (non-breaking change which improves code reliability and robustness)

## Checklist

Before you submit your pull request, please make sure you meet these requirements:

- [ ] Followed the inspiration of the Express.js framework for new functionalities, making them similar in usage.
- [x] Conducted a self-review of the code and provided comments for complex or critical parts.
- [x] Updated the documentation in the `/docs/` directory for [Fiber's documentation](https://docs.gofiber.io/).
- [x] Added or updated unit tests to validate the effectiveness of the changes or new features.
- [x] Ensured that new and existing unit tests pass locally with the changes.
- [ ] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community.
- [x] Aimed for optimal performance with minimal allocations in the new code.
- [ ] Provided benchmarks for the new code to analyze and improve upon.

## Commit formatting

Please use emojis in commit messages for an easy way to identify the purpose or intention of a commit. Check out the emoji cheatsheet here: [CONTRIBUTING.md](https://github.com/gofiber/fiber/blob/master/.github/CONTRIBUTING.md#pull-requests-or-commits)

### CODE REVIEW

File: ctx_test.go [hunk 1]
```diff
@@ -5847,6 +5847,20 @@ func Test_GenericParseTypeBoolean(t *testing.T) {
 	}
 }
 
+// go test -run Test_Ctx_Drop -v
+func Test_Ctx_Drop(t *testing.T) {
+	t.Parallel()
+
+	app := New()
+	app.Get("/block-me", func(c Ctx) error {
```
Comment (gaby) [inline ctx_test.go:5855]: Add another handler here that just calls `return`, and test the output

File: ctx.go [hunk 2]
```diff
@@ -1978,3 +1978,9 @@ func (c *DefaultCtx) setMatched(matched bool) {
 func (c *DefaultCtx) setRoute(route *Route) {
 	c.route = route
 }
+
+func (c *DefaultCtx) Drop() error {
+	conn := c.RequestCtx().Conn()
```
Comment (gaby) [inline ctx.go:1983]: this can be one line

return c.RequestCtx().Conn().Close()

**Review by gaby — CHANGES_REQUESTED** [review]
See comments

File: ctx_test.go (hunk 1 above)
Comment (ryanbekhen) [reply ctx_test.go:5857]: please check again

File: ctx.go (hunk 2 above)
Comment (ryanbekhen) [reply ctx.go]: please check again

**Comment by ryanbekhen** [comment]
@gaby I have pushed the changes. please check.

**Comment by gaby** [comment]
> @gaby I have pushed the changes. please check.

Fix the golangci-lint issue, and it should be good.

**Comment by gaby** [comment]
@grivera64 Added you to reviewers since you did the initial research on this.

**Comment by ryanbekhen** [comment]
@gaby done

**Review by gaby — APPROVED** [review]
👍 LGTM

File: ctx.go [hunk 3]
```diff
@@ -1978,3 +1978,8 @@ func (c *DefaultCtx) setMatched(matched bool) {
 func (c *DefaultCtx) setRoute(route *Route) {
 	c.route = route
 }
+
+func (c *DefaultCtx) Drop() error {
```
Comment (grivera64) [inline ctx.go:1982]: Please add a doc comment here explaining what `Drop()` does. You can use the same description you wrote for the feature in `docs/api/ctx.md`.

File: ctx_test.go [hunk 4]
```diff
@@ -5847,6 +5847,35 @@ func Test_GenericParseTypeBoolean(t *testing.T) {
...
+	// Test the Drop method
+	resp, err := app.Test(httptest.NewRequest(MethodGet, "/block-me", nil))
+	require.Error(t, err)
+	require.Nil(t, resp)
+
+	// Test the no-response handler
+	resp, err = app.Test(httptest.NewRequest(MethodGet, "/no-response", nil))
+	require.NoError(t, err)
+	require.NotNil(t, resp)
+	require.Equal(t, 200, resp.StatusCode)
```
Comment (grivera64) [inline ctx_test.go:5875]: Nit: Please use fiber's `StatusOK` constant here for clarity.

**Review by grivera64 — CHANGES_REQUESTED** [review]
Thanks for helping with this issue!

I just have a few minor comments, please take a look.

**Comment by ReneWerner87** [comment]
@coderabbitai what is the name of the method in expressjs which is the same as this feature

**Comment by ReneWerner87** [comment]
@coderabbitai suggest good method names for this feature

**Comment by ReneWerner87** [comment]
should we use a different name? what do you say @gofiber/maintainers ?

File: ctx.go (hunk 3 above)
Comment (ryanbekhen) [reply ctx.go:1982]: Ok i'll make the changes later

**Review by sixcolors — CHANGES_REQUESTED** [review]
Thank you for the contribution! The `Drop` method is a valuable addition to the `Ctx` interface, and the accompanying documentation and tests are solid. I have a couple of minor suggestions for further improvements:  

1. **Error Handling:**  
   - The `Drop` method includes a `nolint:wrapcheck` directive. Adding a comment to explain why error wrapping is intentionally avoided could be helpful for maintainers and future contributors.  

2. **Test Coverage:**  
   - While the test for Drop is thorough, edge cases like attempting to Drop an already-closed connection could be tested to ensure robust behavior. Also, While I verified that middleware works as expected with `Drop`, it might be beneficial to include explicit test cases verifying middleware behavior when `Drop` is called. This ensures future changes do not inadvertently break compatibility.  

   For example:  
   ```go
   // Middleware that sets a header after the handler is called
   app.Use(func(c Ctx) error {
       err := c.Next()
       c.Response().Header.Set("X-Test", "test")
       return err
   })
   ```

**Comment by sixcolors** [comment]
Good question! Here's a quick comparison of what similar functionality is typically called in other frameworks:

1. **Express.js (Node.js):**
   - No direct equivalent. Developers often use `res.end()` or `req.socket.destroy()` to close connections.

2. **Django (Python):**
   - Django handles connection termination internally; no explicit method is provided.

3. **Spring Boot (Java):**
   - Similarly, connection termination is not exposed as a direct method, and developers manipulate sockets or use HTTP error responses.

4. **ASP.NET Core (C#):**
   - No explicit API for silent connection termination; developers manage sockets manually when needed.

5. **Go Frameworks (e.g., Echo, Gin):**
   - These frameworks do not offer explicit methods for silently terminating client connections either. Termination typically occurs at the socket level (e.g., `net.Conn.Close()`).

### Recommendation:
The name `Drop` feels idiomatic for Go, given its brevity and alignment with networking terminology ("drop the connection").

If we want alternatives for clarity or consistency with other frameworks, we could consider:
- `Abort`
- `Close`
- `CloseSilently`
- `Terminate`
- `TerminateSilently`

Appending "Connection" isn't necessary since this method is part of the connection context.

That said, **`Drop` LGTM**—it succinctly conveys the intent and adheres to Go's minimalist naming conventions.

**Comment by ReneWerner87** [comment]
Ok then let's stick to drop

**Comment by ryanbekhen** [comment]
@sixcolors ok, I'll make the changes later

**Comment by ryanbekhen** [comment]
> > should we use a different name? what do you say @gofiber/maintainers ?
> 
> The name `Drop` feels idiomatic for Go, given its brevity and alignment with networking terminology ("drop the connection").
> 
> If we want alternatives for clarity or consistency with other frameworks, we could consider:
> 
> * `Abort`
> * `Close`
> * `CloseSilently`
> * `Terminate`
> * `TerminateSilently`
> 
> Appending "Connection" isn't necessary since this method is part of the connection context.
> 
> That said, **`Drop` LGTM**—it succinctly conveys the intent and adheres to Go's minimalist naming conventions.

done, please check again.

File: ctx_test.go [hunk 5]
```diff
@@ -5847,6 +5847,63 @@ func Test_GenericParseTypeBoolean(t *testing.T) {
...
+// go test -run Test_Ctx_DropWithMiddleware -v
+func Test_Ctx_DropWithMiddleware(t *testing.T) {
+	t.Parallel()
+
+	app := New()
+
+	// Middleware that calls Drop
+	app.Use(func(c Ctx) error {
+		err := c.Next()
+		c.Response().Header.Set("X-Test", "test")
```
Comment (grivera64) [inline ctx_test.go:5888]: Could we rewrite this as the following for readability?

```go
c.Set("X-Test", "test")
```

File: ctx_test.go [hunk 6]
```diff
@@ -5847,6 +5847,63 @@ func Test_GenericParseTypeBoolean(t *testing.T) {
...
+	// Handler that calls Drop
+	app.Get("/block-me", func(c Ctx) error {
+		return c.Drop()
+	})
+
+	// Test the Drop method
+	resp, err := app.Test(httptest.NewRequest(MethodGet, "/block-me", nil))
+	require.Error(t, err)
+	require.Nil(t, resp)
+	require.Panics(t, func() {
```
Comment (grivera64) [inline ctx_test.go:5901]: I'm not sure if this `require.Panics` is required.

When `resp == nil`, any operation/access done on it should presumably panic due to `invalid memory address or nil pointer dereference`.

I think it should be safe to remove it if we have `require.Nil(t, resp)` above.

File: ctx_test.go (hunk 5 above)
Comment (ryanbekhen) [reply ctx_test.go:5888]: ok thx, i'll change

**Comment by ryanbekhen** [comment]
done @grivera64

**Review by gaby — APPROVED** [review]
LGTM

**Review by efectn — APPROVED** [review]

**Review by grivera64 — APPROVED** [review]
LGTM as well! 🚀

**Review by ReneWerner87 — APPROVED** [review]
